feat(ui-shell): add responsive UX for header nav

jnm2377 opened the pull request:
Closes #2994 
Closes #2995 

Adds functionality for header navigation items to collapse into sidenav.

jnm2377 committed:
feat: create ref

jnm2377 committed:
feat: reference header nav

jnm2377 committed:
feat: render header nav in sidenav

jnm2377 committed:
feat: add header nav-sidenav styles

jnm2377 committed:
feat: update stories

tw15egan commented on packages/react/src/components/UIShell/SideNav.js:
```diff
@@ -45,33 +52,78 @@ const SideNav = React.forwardRef(function SideNav(props, ref) {
+                  const text = i.innerText;
+                  const url = i.children[0].attributes[0].nodeValue;
+                  return (
+                    <SideNavLink href={url} key={`${Math.random()}`}>
```
Since these items won't be reordered, would it be better to add an `index` to the `map` function and use that instead of `Math.random()`? Still not ideal, but it seems like `Math.random()` can cause unnecessary re-renders and is being called each time the menu is opened/closed.

It seems like the best case scenario is if each `SideNavLink` was provided a unique id that we could use as the key. 

For reference: https://stackoverflow.com/questions/29808636/when-giving-unique-keys-to-components-is-it-okay-to-use-math-random-for-gener

asudoh replied:
@jnm2377 A bit related here, any reason behind `<SideNav>` itself rendering children as opposed to letting user render children that we did before...?

tw15egan replied:
I think it is because they are coming directly from the header, as opposed to being created inside the `SideNav`, when the screen goes to mobile

jnm2377 replied:
Yes, @aledavila and I both thought it was better to use the header components as a ref rather than make users write two instances of the same navigation.

jnm2377 replied:
Also, @tw15egan can definitely update it to use the index

tw15egan committed:
Merge branch 'master' into 2994-header-ux

tw15egan committed:
Merge branch 'master' into 2994-header-ux

aagonzales reviewed (commented):
- The `bx--side-nav__submenu` is missing a hover. Should get the same hover as the other items in the menu.
![image](https://user-images.githubusercontent.com/11670886/59790141-d9e52900-9294-11e9-84a1-d26f074622c8.png)

- There should be a page overlay when the menu is open on smaller screens. (use `$overlay-01`)
![image](https://user-images.githubusercontent.com/11670886/59789959-7c50dc80-9294-11e9-8135-8a8aa49c619f.png)

- in the examples can we make the page background Gray 10? So the side nav shows up better (if we're not going to add a border for now) 
![image](https://user-images.githubusercontent.com/11670886/59790741-40b71200-9296-11e9-83d8-625aa68bbbcc.png)

tw15egan committed:
Merge branch 'master' into 2994-header-ux

asudoh commented:
@jnm2377 Thank you for your explanation wrt my question wrt `SideNav.js` change: https://github.com/carbon-design-system/carbon/pull/3105#discussion_r295077105

The biggest thing I'd like to see happening in this PR is the capability of [rendering side nav items from `<HeaderNavigation>` content](https://github.com/carbon-design-system/carbon/blob/88d6d5eeb22a35a81fb0e23161f7d0433679ae25/packages/react/src/components/UIShell/SideNav.js#L90-L123) won't be baked into `<SideNav>`. Given the intent seems avoiding duplicate list (of React nodes) application needs to maintain, I have suggestions with two choices:

1. Relegating the capability of maintaining the children of `<HeaderNavigation>` to `<HeaderContainer>` (Example: https://github.com/asudoh/carbon-components/commit/b08708f68a97f2b87e6c6d74433fe9183b728f03)
2. Letting application simply render contents of `<HeaderNavigation>` and `<SideNavItems>` from a single data - Which means application simply has the map logic like https://github.com/asudoh/carbon-components/blob/b08708f68a97f2b87e6c6d74433fe9183b728f03/packages/react/src/components/UIShell/UIShell-story.js#L255-L271 in both of `<HeaderNavigation>` and `<SideNavItems>` rendering code in applications.

I'm preferring 2. as I consider several edge cases like `<HeaderMenuItem>` with non-text `children`, `<HeaderMenu>` with non-text `menuLinkName` as well as for the sake of avoiding [children-updated detection logic in `<HeaderNavigation>`](https://github.com/asudoh/carbon-components/blob/b08708f68a97f2b87e6c6d74433fe9183b728f03/packages/react/src/components/UIShell/HeaderNavigation.js#L56-L68). If we can take 2., we can introduce a helper for [the map logic](https://github.com/asudoh/carbon-components/blob/b08708f68a97f2b87e6c6d74433fe9183b728f03/packages/react/src/components/UIShell/UIShell-story.js#L255-L271) to make it easier for applications.

emyarod reviewed (commented):
are the commented lines to be removed or still WIP?

tw15egan commented:
@emyarod I believe the rail will be added back in in a future update, so the comments are just temporary

emyarod commented on packages/react/src/components/UIShell/HeaderContainer.js:
```diff
@@ -10,10 +14,13 @@ const HeaderContainer = ({ isSideNavExpanded, render: Children }) => {
     setIsSideNavExpandedState(!isSideNavExpandedState);
   }, [isSideNavExpandedState, setIsSideNavExpandedState]);
 
+  headerNavRef = React.createRef();
```
does the `ref` need to persist between renders? we can't use `createRef` in a function component right?

tw15egan committed:
Merge branch 'master' into 2994-header-ux

jnm2377 commented:
@asudoh option 2 seems like a viable option. I'll work on refactoring this. 👍

tw15egan committed:
Merge branch 'master' into 2994-header-ux

tw15egan committed:
Merge branch 'master' into 2994-header-ux

tw15egan committed:
Merge branch 'master' into 2994-header-ux

tw15egan committed:
Merge branch 'master' into 2994-header-ux

jnm2377 committed:
feat: create header sidenav item

jnm2377 committed:
feat: remove header items from sidenav

jnm2377 committed:
feat: remove header nav ref

jnm2377 committed:
feat: add HeaderSideNavItems

jnm2377 committed:
feat: add overlay and header sidenav styles

jnm2377 committed:
feat: update story

jnm2377 committed:
Merge branch '2994-header-ux' of github.com:jnm2377/carbon-components into 2994-header-ux

jnm2377 commented:
Update for dev reviewers: @tw15egan @emyarod @aledavila @asudoh rather than creating a ref and mapping through items, we decided to just allow `HeaderSideNavItems` to render  any `{children}` like its sibling components do. Users then have 3 ways of including header links in the sidenav:

1. copy and paste the exact header nav components into `HeaderSideNavItems` (I added overriding styles for those components so that they look like sidenav components when they're nested inside of `bx--side-nav`

```css
<SideNav
   aria-label="Side navigation"
   expanded={isSideNavExpanded}
   isPersistent={false}>
      <SideNavItems>
         <HeaderSideNavItems>
               <HeaderMenuItem href="">Link 1</HeaderMenuItem>
               <HeaderMenuItem href="">Link 2</HeaderMenuItem>
               <HeaderMenuItem href="">Link 3</HeaderMenuItem>
               <HeaderMenu aria-label="Link 4" menuLinkName="Link 4">
                   <HeaderMenuItem href="">Sub-link 1</HeaderMenuItem>
                   <HeaderMenuItem href="">Sub-link 2</HeaderMenuItem>
                   <HeaderMenuItem href="">Sub-link 3</HeaderMenuItem>
                </HeaderMenu>
           </HeaderSideNavItems>
      </SideNavItems>
</SideNav>
```
2. they can use the actual side nav components inside of `HeaderSideNavItems` and just re-type the url/link names

```css
<SideNav
   aria-label="Side navigation"
   expanded={isSideNavExpanded}
   isPersistent={false}>
      <SideNavItems>
         <HeaderSideNavItems>
               <SideNavMenuItem href="">Link 1</SideNavMenuItem >
               <SideNavMenuItem  href="">Link 2</SideNavMenuItem >
               <SideNavMenuItem  href="">Link 3</SideNavMenuItem >
               <SideNavMenu title="Link 4">
                   <SideNavMenuItem  href="">Sub-link 1</SideNavMenuItem >
                   <SideNavMenuItem  href="">Sub-link 2</SideNavMenuItem >
                   <SideNavMenuItem  href="">Sub-link 3</SideNavMenuItem >
                </SideNavMenu>
           </HeaderSideNavItems>
      </SideNavItems>
</SideNav>
```
3. they can create an object with the url/link names and map through it to render header items and sidenav items with the same data

We decided that having these simple versions would be a lot more user friendly than what I or Akira had initially.

Update for design review: @aagonzales I added the overlay, and hover styles, as well as styled the main content bg color. The reason the main content bg color isn't taking up the full width is actually because Storybook itself is adding padding to the demo so there's no way of overriding that (you can't tell from the header or sidenav bc those have fixed positioning but it's obvious on the main content).

jnm2377 committed:
Merge branch '2994-header-ux' of github.com:jnm2377/carbon-components into 2994-header-ux

jnm2377 committed:
chore: update tests

jnm2377 committed:
feat: fix header menu sidenav icon

tw15egan committed:
Merge branch 'master' into 2994-header-ux

asudoh reviewed (approved):
Your decision makes sense to me - Thanks @jnm2377 for thinking it through!

tw15egan committed:
Merge branch 'master' into 2994-header-ux

tw15egan reviewed (approved):
This looks great! I'm glad we finally settled on this approach 🎉

tw15egan commented:
Just need @aagonzales or @designertyler to sign off and we're good to merge 👍

aagonzales commented:
If a sub-menu has a selected item in it and the sub-menu is closed then the parent item should also get the selected state style. Like what we do on our website, example below:

![Jul-08-2019 11-23-12](https://user-images.githubusercontent.com/11670886/60826213-fdeca980-a172-11e9-8687-987218f7b635.gif)

tw15egan committed:
Merge branch 'master' into 2994-header-ux

jnm2377 committed:
Merge branch 'master' into 2994-header-ux

jnm2377 committed:
fix: update story to show active sidenav menu

jnm2377 committed:
fix: remove overlay if fixed sidenav

aagonzales reviewed (approved):
Perfect! Thanks for updating.
